### DISCUSSION TITLE
Is it possible to define custom commands that can take arguments?

### DISCUSSION THREAD

**Discussion by cadadr** [qutebrowser/qutebrowser#5657] [Q&A] [answered]
Hi all!

Maybe I'm missing something obvious here---sorry if that's the case---but is it possible to define a command alias or a command that takes arbitrary arguments in config.py?

**Comment by toofar** [comment, marked as answer]
You can import cmdutils and register a command like so:
```
from qutebrowser.api import cmdutils

@cmdutils.register()
def my_command(some_flag: bool = False):
  pass
```

**Reply by cadadr** [reply]
Thanks! I recall reading somewhere in the docs that importing qutebrowser stuff can be harmful. Is this a safe thing to do, or am I better of sticking to what `c` and `config` have to offer?

**Reply by toofar** [reply]
It is mostly harmful in that users who do weird things with a projects internals tend to take up an inordinate amount of a maintainer's time in support. Although I have learned to back up my basedir regularly because if you make it crash in new and exciting ways the existing recovery mechanisms might not know how to handle it (losing my session has happened before).
As I understand it everything under qutebrowser.api is meant for the as yet unfinished [extension api](https://github.com/qutebrowser/qutebrowser/issues/30) so it is intended to be used by third parties, even if it isn't finalised yet. cmdutils is the module used to register all the first party commands as well so it should be pretty stable.
Anyway if you think you would get value out of customizing beyond what is documented I say go for it.

**Reply by cadadr** [reply]
Thanks! Really appreciate your detailed replies!

**Reply by The-Compiler** [reply]
I don't have much more to add to that - doing that kind of thing isn't official yet and the API might change at some point, but it's been pretty stable so far. As long as you're okay with keeping both pieces when something breaks, go for it :slightly_smiling_face:
